fasm/__init__.py

```diff
@@ -46,122 +60,134 @@ class ValueFormat(enum.Enum):
-            address_width = end - start + 1
-        else:
-            start = int(set_feature_model.feature_address.address1, 10)
-            end = None
-            address_width = 1
+    feature, p = tagged_string_from_bytes('f', data, i)
+    start, end, p = fasm_address_from_bytes(data, p)
+    value, value_format, p = fasm_value_from_bytes(data, p)
 
-    if set_feature_model.feature_value:
-        width, value, value_format = verilog_value_to_int(
-            set_feature_model.feature_value)
+    if not value:
+        value = 1
 
-        if width is not None:
-            assert width <= address_width
-
-        assert value < (2**address_width), (value, address_width)
+    if end and not start:
+        start = end
 
     return SetFasmFeature(
-        feature=set_feature_model.feature,
-        start=start,
-        end=end,
-        value=value,
-        value_format=value_format,
-    )
-
-
-def get_fasm_metamodel():
-    return textx.metamodel_from_file(
-        file_name=os.path.join(os.path.dirname(__file__), 'fasm.tx'),
-        skipws=False)
-
-
-def fasm_model_to_tuple(fasm_model):
-    """ Converts FasmFile model to list of FasmLine named tuples. """
-    if not fasm_model:
-        return
-
-    for fasm_line in fasm_model.lines:
-        set_feature = None
-        annotations = None
-        comment = None
-
-        if fasm_line.set_feature:
-            set_feature = set_feature_model_to_tuple(fasm_line.set_feature)
-
-        if fasm_line.annotations:
-            annotations = tuple(
-                Annotation(
-                    name=annotation.name,
-                    value=annotation.value if annotation.value else '')
-                for annotation in fasm_line.annotations.annotations)
-
-        if fasm_line.comment:
-            comment = fasm_line.comment.comment
-
-        yield FasmLine(
-            set_feature=set_feature,
-            annotations=annotations,
-            comment=comment,
-        )
-
+        feature = feature,
+        start = start,
+        end = end,
+        value = value,
+        value_format = value_format
+    ), i+length
+
+# Decode an annotation: x = "y"
+def fasm_annotation_from_bytes(data, i):
+    length, i = get_header('a', data, i)
+    if not length:
+        return None, i
+
+    name, p = tagged_string_from_bytes('.', data, i)
+    value, p = tagged_string_from_bytes('=', data, p)
+    return Annotation(
+        name = name,
+        value = value
+    ), i+length
+
+# Decode a set of annotations: { ... }
+def fasm_annotations_from_bytes(data, i):
+    length, i = get_header('{', data, i)
+    if not length:
+        return None, i
+
+    annotations = []
+    annotation, p = fasm_annotation_from_bytes(data, i)
+    while annotation:
+        annotations.append(annotation)
+        annotation, p = fasm_annotation_from_bytes(data, p)
+
+    return annotations, i+length
+
+# Decode an entire FASM line
```

Comment:
Rather than comments above the function, these should be docstrings?  Please following something like https://numpydoc.readthedocs.io/en/latest/format.html

Reply:
Done.